File: examples/delphi/factory_method.pas
{$mode objfpc}{$H+}
program factory_method;

uses SysUtils;

type
  TTransport = class(TObject)
  public
    function Deliver: string; virtual; abstract;
  end;

  TTruck = class(TTransport)
  public
    function Deliver: string; override;
  end;

  TShip = class(TTransport)
  public
    function Deliver: string; override;
  end;

  TLogistics = class(TObject)
  public
    function CreateTransport: TTransport; virtual; abstract;
    procedure PlanDelivery;
  end;

  TRoadLogistics = class(TLogistics)
  public
    function CreateTransport: TTransport; override;
  end;

  TSeaLogistics = class(TLogistics)
  public
    function CreateTransport: TTransport; override;
  end;

function TTruck.Deliver: string;
begin
  Result := 'Delivering by land in a truck';
end;

function TShip.Deliver: string;
begin
  Result := 'Delivering by sea in a ship';
end;

function TRoadLogistics.CreateTransport: TTransport;
begin
  Result := TTruck.Create;
end;

function TSeaLogistics.CreateTransport: TTransport;
begin
  Result := TShip.Create;
end;

procedure TLogistics.PlanDelivery;
var
  Transport: TTransport;
begin
  Transport := CreateTransport;
  WriteLn(Transport.Deliver);
  Transport.Free;
end;

var
  Logistics: TLogistics;
begin
  WriteLn('=== Factory Method Pattern: Transport Logistics ===');
  WriteLn;

  WriteLn('Road logistics:');
  Logistics := TRoadLogistics.Create;
  Logistics.PlanDelivery;
  Logistics.Free;

  WriteLn('Sea logistics:');
  Logistics := TSeaLogistics.Create;
  Logistics.PlanDelivery;
  Logistics.Free;
end.

File: examples/cpp/factory-method.cpp
#include <iostream>
#include <memory>
#include <string>

// Product interface
class Transport {
public:
    virtual ~Transport() = default;
    virtual std::string deliver() const = 0;
};

class Truck : public Transport {
public:
    std::string deliver() const override {
        return "Delivering by land in a truck";
    }
};

class Ship : public Transport {
public:
    std::string deliver() const override {
        return "Delivering by sea in a ship";
    }
};

// Creator
class Logistics {
public:
    virtual ~Logistics() = default;
    virtual std::unique_ptr<Transport> createTransport() const = 0;

    std::string planDelivery() const {
        auto transport = createTransport();
        return "Logistics planned: " + transport->deliver();
    }
};

class RoadLogistics : public Logistics {
public:
    std::unique_ptr<Transport> createTransport() const override {
        return std::make_unique<Truck>();
    }
};

class SeaLogistics : public Logistics {
public:
    std::unique_ptr<Transport> createTransport() const override {
        return std::make_unique<Ship>();
    }
};

int main() {
    std::cout << "=== Factory Method: Transport Logistics ===" << std::endl;

    std::unique_ptr<Logistics> road = std::make_unique<RoadLogistics>();
    std::cout << road->planDelivery() << std::endl;

    std::unique_ptr<Logistics> sea = std::make_unique<SeaLogistics>();
    std::cout << sea->planDelivery() << std::endl;

    return 0;
}
